x/incentives/keeper/pot.go

```diff
@@ -0,0 +1,308 @@
+	pot, err := k.GetPotByID(ctx, potID)
+	if err != nil {
+		return err
+	}
+	if err := k.bk.SendCoinsFromAccountToModule(ctx, owner, types.ModuleName, coins); err != nil {
+		return err
+	}
+
+	pot.Coins = pot.Coins.Add(coins...)
+	k.setPot(ctx, pot)
+	return nil
+}
+
+// BeginDistribution is a utility to begin distribution for a specific pot
+func (k Keeper) BeginDistribution(ctx sdk.Context, pot types.Pot) error {
+	// validation for current time and distribution start time
+	curTime := ctx.BlockTime()
+	if curTime.Before(pot.StartTime) {
+		return fmt.Errorf("pot is not able to start distribution yet: %s >= %s", curTime.String(), pot.StartTime.String())
+	}
+
+	timeKey := getTimeKey(pot.StartTime)
+	k.deletePotRefByKey(ctx, combineKeys(types.KeyPrefixUpcomingPots, timeKey), pot.Id)
+	k.addPotRefByKey(ctx, combineKeys(types.KeyPrefixActivePots, timeKey), pot.Id)
+	return nil
+}
+
+// FinishDistribution is a utility to finish distribution for a specific pot
+func (k Keeper) FinishDistribution(ctx sdk.Context, pot types.Pot) error {
+	timeKey := getTimeKey(pot.StartTime)
+	k.deletePotRefByKey(ctx, combineKeys(types.KeyPrefixActivePots, timeKey), pot.Id)
+	k.addPotRefByKey(ctx, combineKeys(types.KeyPrefixFinishedPots, timeKey), pot.Id)
+	return nil
+}
+
+// GetLocksToDistribution get locks that are associated to a condition
+func (k Keeper) GetLocksToDistribution(ctx sdk.Context, distrTo types.DistrCondition) []lockuptypes.PeriodLock {
+	switch distrTo.LockQueryType {
+	case types.ByDuration:
+		return k.lk.GetLocksLongerThanDurationDenom(ctx, distrTo.Denom, distrTo.Duration)
+	case types.ByTime:
+		return k.lk.GetLocksPastTimeDenom(ctx, distrTo.Denom, distrTo.Timestamp)
+	default:
+	}
+	return []lockuptypes.PeriodLock{}
+}
+
+// FilteredLocksDistributionEst estimate distribution amount coins from pot for fitting conditions
+func (k Keeper) FilteredLocksDistributionEst(ctx sdk.Context, pot types.Pot, filteredLocks []lockuptypes.PeriodLock) (types.Pot, sdk.Coins, error) {
+	filteredLockIDs := make(map[uint64]bool)
+	for _, lock := range filteredLocks {
+		filteredLockIDs[lock.ID] = true
+	}
+
+	totalDistrCoins := sdk.NewCoins()
+	filteredDistrCoins := sdk.NewCoins()
+	locks := k.GetLocksToDistribution(ctx, pot.DistributeTo)
+	lockSum := lockuptypes.SumLocksByDenom(locks, pot.DistributeTo.Denom)
+
+	if lockSum.IsZero() {
+		return types.Pot{}, nil, nil
+	}
+
+	remainCoins := pot.Coins.Sub(pot.DistributedCoins)
+	remainEpochs := pot.NumEpochs - pot.FilledEpochs
+	for _, lock := range locks {
+		distrCoins := sdk.Coins{}
+		for _, coin := range remainCoins {
+			amt := coin.Amount.Mul(lock.Coins.AmountOf(pot.DistributeTo.Denom)).Quo(lockSum.Mul(sdk.NewInt(int64(remainEpochs))))
+			if amt.IsPositive() {
+				distrCoins = distrCoins.Add(sdk.NewCoin(coin.Denom, amt))
+			}
+		}
+		distrCoins = distrCoins.Sort()
+		if !distrCoins.Empty() && (len(filteredLocks) == 0 || filteredLockIDs[lock.ID]) {
```

Comment:
Lets change the loop so we skip computing `distrCoins` if `!(len(filteredLocks) == 0 || filteredLockIDs[lock.ID])`. This is of high computational complexity for nodes even with this.

Reply:
Ok

Reply:
we need to calculate both totalDistrCoins and filteredDistrCoins, and I think there's no other way to simplify this.